File: backend/internal/handler/problem.go
package handler

import (
	"context"
	"encoding/json"
	"log/slog"
	"net/http"

	"github.com/Natti3588/Ippo/backend/internal/api"
)

// writeProblem は RFC 9457 の Problem Details を書き出す。
//
// detail に渡してよいのは、利用者に見せて差し支えない説明だけである。
// err.Error() をそのまま渡してはならない。内部の文言やラップされた情報が外に出る。
//
// 500 はこの関数を使わない。契約に 500 を宣言していないため、本文を持たせない。
func writeProblem(ctx context.Context, w http.ResponseWriter, logger *slog.Logger, status int, detail string) {
	w.Header().Set("Content-Type", "application/problem+json")
	w.WriteHeader(status)

	body := api.ProblemDetails{
		Type:   "about:blank",
		Title:  http.StatusText(status),
		Status: int32(status),
		Detail: detail,
	}
	if err := json.NewEncoder(w).Encode(body); err != nil {
		logger.ErrorContext(ctx, "エラーレスポンスの書き込みに失敗", "error", err)
	}
}

File: frontend/src/lib/problem.ts
import type { components } from "./api-types";

/** 契約で定義した RFC 9457 の Problem Details。 */
export type ProblemDetails = components["schemas"]["ProblemDetails"];

/**
 * API がエラーを返したことを表す。
 *
 * detail に入るのは、利用者に見せてよい日本語だけ（バックエンドがそう作ってある）。
 * そのまま画面に出せる。title は HTTP のステータス文なので出さない。
 */
export class ApiError extends Error {
  readonly status: number;
  readonly detail: string;

  constructor(status: number, detail: string) {
    super(detail);
    this.name = "ApiError";
    this.status = status;
    this.detail = detail;
  }
}

/**
 * レスポンスから ApiError を作る。
 *
 * 500 は契約に書いていないので本文が無い。Problem Details として読めなければ、
 * ここで汎用の文言に倒す。画面側で毎回「detail が空かもしれない」を
 * 考えずに済ませたい。
 */
export async function toApiError(res: Response): Promise<ApiError> {
  const fallback = "通信に失敗しました。しばらくしてからお試しください";

  try {
    const body = (await res.json()) as ProblemDetails;
    const detail = body.detail?.trim();
    return new ApiError(res.status, detail ? detail : fallback);
  } catch {
    return new ApiError(res.status, fallback);
  }
}
